### ISSUE TITLE
GPU: Bundle_Adjuster of Colmap 311

### ISSUE THREAD

**Issue by HungNgoCT** [colmap/colmap#2840]
Hi,

I saw that there is an option:
--BundleAdjustment.use_gpu (0/1)
for bundle_adjuster. But when I set 1, it seems bundle adjustment does not run on GPU.

I choose correct GPU index, and GPU is used for extract features and matching

![image](https://github.com/user-attachments/assets/029c8bac-257d-4f42-861b-b7f2919c2f49)

I am wondering if Colmap 311 supports BA using GPU or not?

Thank you

**Comment by ahojnnes** [comment]
For now, this requires installing the latest ceres dev version with GPU support (CUDA and cuDSS).

**Comment by sarlinpe** [comment]
Support for cuDSS in Ceres was added in August, long after Ceres 2.2.0 was released. It seems like 2.3.0 will be released soon.

**Comment by mattiasmar** [comment]
I see. However, when taking the latest commit to ceres master I encounter the below. 
Is there a dockerfile ready that encapsulates the colmap build with cudSS support?

```
...
-- Found CUDA version 12.4.131 installed in: /usr/local/cuda
-- The CUDA compiler identification is NVIDIA 12.4.131
-- Detecting CUDA compiler ABI info
-- Detecting CUDA compiler ABI info - done
-- Check for working CUDA compiler: /usr/local/cuda/bin/nvcc - skipped
-- Detecting CUDA compile features
-- Detecting CUDA compile features - done
-- Setting CUDA Architecture to 50;60;70;75;80;90
CMake Error at /usr/lib/x86_64-linux-gnu/cmake/cudss/cudss-targets.cmake:94 (message):
  The imported target "cudss" references the file

     "/usr/lib/x86_64-linux-gnu/./libcudss.so.0.3.0"

  but this file does not exist.  Possible reasons include:

  * The file was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and contained

     "/usr/lib/x86_64-linux-gnu/cmake/cudss/cudss-targets.cmake"

  but not all the files it references.

Call Stack (most recent call first):
  /usr/lib/x86_64-linux-gnu/cmake/cudss/cudss-config.cmake:134 (include)
  CMakeLists.txt:355 (find_package)


-- Configuring incomplete, errors occurred!

```

**Comment by mattiasmar** [comment]
@sarlinpe Is there a default installation of colmap that also brings a version of ceres with cudSS support?

**Comment by ahojnnes** [comment]
@mattiasmar There is no existing docker image that supports this. You'll have to manually install it for now. cuDSS is not part of the standard CUDA distribution. You have to manually download and install it from: https://developer.nvidia.com/cudss

**Comment by mattiasmar** [comment]
@ahojnnes  Thanks. So I figured and created the below Dockerfile prior to posting here. Even though it runs and indicates that it was compiled against CUDA (and I can also see that some GPU resources gets in use during runtime), the bundle adjuster still complains about ceres not being compiled with cudSS support. Note that I use the latest commit of ceres (c29b5257e23f91d6a47c4db9d57350ed4985ea46, from end of September this year). Can you see what I'm missing? 
I'm on a Titan V which use SM architecture "70", which ought to be supported by [cudSS](https://docs.nvidia.com/cuda/cudss/index.html) .

`ARG UBUNTU_VERSION=22.04
ARG NVIDIA_CUDA_VERSION=12.4.1

ARG COLMAP_GIT_COMMIT=91840e63f30ac04c2ac7460da5b60df9fed38a34
ARG CUDA_ARCHITECTURES=70
ENV QT_XCB_GL_INTEGRATION=xcb_egl

# Prevent stop building ubuntu at time zone selection.
ENV DEBIAN_FRONTEND=noninteractive

# Prepare and empty machine for building.
RUN apt-get update && \
    apt-get install -y --no-install-recommends --no-install-suggests \
    wget \ 
    vim \
    tmux \
    qtbase5-dev \
    python3.10-venv \
    python3-pip \
    python3 \
    nvidia-cuda-toolkit-gcc- \
    ninja-build \
    nano \ 
    libtbb-dev \
    libsuitesparse-dev \
    libsqlite3-dev \
    libqt5svg5-dev \
    libqt5opengl5-dev \
    libmetis-dev \
    libgtest-dev \
    libgoogle-glog-dev \
    libgmock-dev \
    libglew-dev \
    libgflags-dev \
    libfreeimage-dev \
    libflann-dev \
    libeigen3-dev \
    libcgal-dev \
    libboost-system-dev \
    libboost-program-options-dev \
    libboost-graph-dev \
    libboost-filesystem-dev \
    libboost-all-dev \
    libatlas-base-dev \
    git \
    cmake \
    build-essential 
    
WORKDIR /ceres
RUN wget https://developer.download.nvidia.com/compute/cuda/repos/ubuntu2204/x86_64/cuda-keyring_1.1-1_all.deb
RUN  dpkg -i cuda-keyring_1.1-1_all.deb
RUN  apt-get update && apt-get -y install cudss

RUN apt-get update && \
    apt-get install -y --no-install-recommends libtbb-dev


RUN git clone --recurse-submodules https://github.com/ceres-solver/ceres-solver && \
    cd ceres-solver && \
    mkdir build && \
    cd build && \
    cmake .. -DBUILD_TESTING=ON -DBUILD_EXAMPLES=ON -DCMAKE_CUDA_ARCHITECTURES=70 && \
    make  && make install

# Build and install COLMAP.
WORKDIR /colmap
RUN git clone  --recurse-submodules https://github.com/colmap/colmap.git &&  cd colmap && \
    git fetch  https://github.com/colmap/colmap.git ${COLMAP_GIT_COMMIT} && \
    git checkout FETCH_HEAD && \
    mkdir build && \
    cd build 
RUN cd colmap/build &&  pwd && cmake .. -DCMAKE_BUILD_TYPE=Release -DBUILD_TESTING=OFF -DCOLMAP_CUDA=ON -DCMAKE_CUDA_ARCHITECTURES=70 && make`

**Comment by glorioushonor** [comment]
> I submitted this [issue](https://github.com/ceres-solver/ceres-solver/issues/1125) to the Ceres-solver forum. It contains my Dockerfile that illustrates the problem.

Hello, I would like to know how much the mapper speed of colmap is improved after using GPU in Bundle Adjustment successfully. If it raises a lot, I would like to try your docker instruction.

**Comment by mattiasmar** [comment]
@glorioushonor  I didn't see any immediate improvement. There are a bunch of heuristic thresholds that one can try to adjust though [Heuristic thresholds to switch from CPU to GPU usage](https://github.com/colmap/colmap/blob/e778be634029abbe14a3470dd4df6e53f475111f/src/colmap/estimators/bundle_adjustment.h#L140-L156). 
The right parameters depends on the numerical _size_ of your problem.

**Comment by ahojnnes** [comment]
My original experiments showed quite significant speedups, especially on larger datasets. @mattiasmar Which GPU do you have?

**Comment by mattiasmar** [comment]
@ahojnnes Titan V

**Comment by ahojnnes** [comment]
I guess you do see some GPU utilization now after compiling correctly? If you pass `--log_level 2` to colmap you should also see that the CUDA solvers are used in the ceres report.

**Comment by mattiasmar** [comment]
yes, I do see that something is loaded to the GPU. I just don't see that processing time was improved.

I didn't run a benchmark. I could if you would point me to some. I also didn't invest time in tuning any of the heuristic cpu/gpu parameters.  The test scene that I'm using has some 8000+ images. Images captured about 50-100cm apart.

**Comment by ahojnnes** [comment]
The Titan V in theory has sufficient double precision units but beating 64 logical Xeon cores may still be hard. Another guess is that the co-visibility pattern in your dataset may lead to a BA problem structure that is not so suitable for benefiting from GPU/CUDA/cuDSS speedups. I am going to stop guessing here, as I have not spent enough time on this topic and only used the Ceres feature from a user perspective.

@S-o-T You may have some insights on this topic, as I believe you originally implemented some of the CUDA functionality in Ceres.

**Comment by S-o-T** [comment]
An estimate of speedup of global BA that could be achieved using cuDSS as ceres's linear solver backend could be inferred from following evaluation that i performed using BAL dataset https://github.com/ceres-solver/ceres-solver/issues/759#issuecomment-2097057662 . As @ahojnnes suggested, the problem's size/structure is significant factor to be considered, in addition to your system characteristics (CPU compute, CPU/RAM bandwidth, RAM/VRAM bandwidth, VRAM/SM bandwidth, SM compute).

Also, current default threshold to switch from direct linear solver (which can leverage cuDSS) to iterative solver (which still bound to CPU only) is 1k images and colmap's quite conservatives defaults to trigger global BA will probably manifest into your 8k images dataset reconstruction time to be dominated by global BA solves after first 1k images (such solves are performed via PCG, not accelerated via cuDSS on GPU). This hypothesis could be quite cheaply tested by inspecting the logs and, e.g., plotting num_registered_images vs wall_clock_time or num_registered_images vs global_BA_wall_clock_time (such metrics are extractable from logs using grep/awk).

If you could provide your reconstruction snapshot before it reaches 1k images, i could try to find some time in order to evaluate/profile performance of BA with cuDSS linear solver on my system.

**Comment by ahojnnes** [comment]
This question has been addressed in #3163. GPU BA requires Ceres compiled with cuDSS support (>= 2.3.0) and only activates above certain reconstruction size thresholds. Precompiled binaries may not include GPU BA support. Closing as answered.
